src/number_theory/modular_forms/modular_forms.lean

```diff
@@ -0,0 +1,391 @@
+  (transf : f ∈ weakly_modular_submodule k Γ)
+  (infinity : ∀ (A : SL(2,ℤ)), is_zero_at_inf (f ∣[k] A))
+
+/-- The zero modular form is a cusp form-/
+lemma zero_cusp_form : is_cusp_form_of_weight_and_level k Γ 0 :=
+{ hol := by {apply mdifferentiable_zero,},
+  transf := (weakly_modular_submodule k Γ).zero_mem',
+  infinity := by { intro A,
+    convert zero_form_is_zero_at_inf,
+    rw slash_k,
+    simp only [pi.zero_apply, zero_mul],
+    refl, } }
+
+lemma is_modular_form_of_weight_and_level_of_is_cusp_form_of_weight_and_level {f : ℍ → ℂ}
+  (h : is_cusp_form_of_weight_and_level k Γ f) : is_modular_form_of_weight_and_level k Γ f :=
+{ hol := h.1,
+  transf := h.2,
+  infinity := λ (A : SL(2,ℤ)), is_zero_at_inf_is_bound _ (h.3 A), }
+
+ /-- The zero modular form is a modular form-/
+lemma zero_mod_form : is_modular_form_of_weight_and_level k Γ 0 :=
+begin
+  apply_rules [is_modular_form_of_weight_and_level_of_is_cusp_form_of_weight_and_level,
+    zero_cusp_form],
+end
+
+/-- This is the space of modular forms of weight `k` and level `Γ`-/
+def space_of_mod_forms_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
+{ carrier := { f : ℍ → ℂ | is_modular_form_of_weight_and_level k Γ f},
+  zero_mem':= by { simp only [set.mem_set_of_eq], apply zero_mod_form, },
+  add_mem' := by { intros a b ha hb,
+    split,
+    exact mdifferentiable_add _ _ ha.hol hb.hol,
+    exact (weakly_modular_submodule k Γ).add_mem' ha.transf hb.transf,
+    intro A, rw slash_k_add,
+    exact (bounded_at_infty_submodule.add_mem' (ha.infinity A) (hb.infinity A)) },
+  smul_mem' := by { intros c f hf,
+    split,
+    exact mdifferentiable_smul _ _ hf.hol,
+    exact (weakly_modular_submodule k Γ).smul_mem' _ hf.transf,
+    intro A,
+    rw smul_slash_k,
+    apply (bounded_at_infty_submodule.smul_mem' c (hf.infinity A)), }, }
+
+localized "notation `M(`k`, `Γ`)`:= space_of_mod_forms_of_weight_and_level k Γ" in modular_forms
+
+/-- This is the space of cuspforms of weigth `k` and level `Γ` -/
+def space_of_cusp_forms_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
+{ carrier := is_cusp_form_of_weight_and_level k Γ,
+  zero_mem' := by apply zero_cusp_form,
+  add_mem' := by { intros a b ha hb,
+    split,
+    exact mdifferentiable_add _ _ ha.hol hb.hol,
+    exact (weakly_modular_submodule k Γ).add_mem' ha.transf hb.transf,
+    intro A, rw slash_k_add,
+    apply (zero_at_infty_submodule.add_mem' (ha.infinity A) (hb.infinity A)) },
+  smul_mem' := by { intros c f hf, split,
+    exact mdifferentiable_smul _ _ hf.hol,
+    exact (weakly_modular_submodule k Γ).smul_mem' _ hf.transf,
+    intro A,
+    rw smul_slash_k,
+    apply zero_at_infty_submodule.smul_mem' c (hf.infinity A), }, }
+
+localized "notation `S(`k`, `Γ`)`:= space_of_cusp_forms_of_weight_and_level k Γ" in modular_forms
+
+/--The product of two modular forms is a modular form whose weight is the sum of the weights-/
```

Comment:
It is a [graded_algebra](https://leanprover-community.github.io/mathlib_docs/ring_theory/graded_algebra/basic.html#graded_algebra) I guess.